SoC-2019-Ideas.md

```diff
@@ -218,6 +218,56 @@ This would consist in taking care of the following issues:
   `This reverts commit <sha-of-rewritten-A>`.)
   Do this for both sha1sums and sha1sum abbreviations in commit messages.
 
+### `git revert --drop` and `git commit --reword`
+
+ - Language: C, shell (bash)
+ - Difficulty: easy
+ - Possible mentors: N/A
+
+The interactive rebase already supports the special oneline prefixes
+`fixup!` and `squash!` in the `--autosquash` mode; these commits will
+be reordered in the todo list and their `pick` commands adjusted
+accordingly.
+
+These commits can be crafted conveniently via the `--fixup` and
+`--squash` options of `git commit`.
+
+The idea of this project is to add two more actions, `drop!` and
+`reword!`:
+
+* The `drop!` action (for convenience, `git revert --drop <commit>`)
+  would not only revert the commit, but a subsequent `git rebase -i
+  --autosquash` would reorder the `drop!` commit directly after the
+  matching commit, then change the matching commit's `pick` to `drop`
+  and comment out the `drop!` one.
+
+* The `reword!` action (for convenience, `git commit --reword <commit>`)
+  would let the user edit the commit message of the referenced commit,
+  and add an "empty" commit (i.e. a commit that does not modify any
+  files) with that commit message, prefixed with the `reword!` oneline.
+  The next `git rebase -i --autosquash` would then not only reorder
+  that commit after verifying that it is indeed an empty commit, it
+  would then also replace the `pick` command with an appropriate new
+  command (say, by extending the `squash` command to accept a `--reword`
+  option).
+
+This project will need to begin by implementing test cases to define
+the expected outcome, and then implement the actual functionality.
+
+### Teach `git stash` to handle unmerged index entries
```

Comment:
As these projects have "Possible mentors: N/A" I think students are less likely to select them. So we can probably keep them in the list. Or maybe we should state clearly somewhere that projects that don't have possible mentors should not be selected.

Reply:
> for fears of stash in C not being ready to have stuff being built on top of it before GSoC starts.

It is in `next`, and developers can build on it. Even if any such patches won't advance beyond where `ps/stash-in-c` happens to be at the moment, that should be fine, no?

> There's still some chance that a student would have to make the changes in both the shell version and the C version, but dunno if we should hold this project back just for that reason.

We now already have that precedent in the built-in rebase that we implemented a feature only in the C code, and simply forced that in the test case by setting `*_USE_BUILTIN=true` inside the test case.

> [...] maybe we should state clearly somewhere that projects that don't have possible mentors should not be selected.

Why don't you live up to the challenge? After all, *you* suggested that students should be challenged, and I don't see why the buck should stop short of the mentors.

Reply:
> It is in next, and developers can build on it. Even if any such patches won't advance beyond where ps/stash-in-c happens to be at the moment, that should be fine, no?

Yes, that's fine.  I should have clarified that I removed it before stash in C was merged into next, but now that it is in next I don't see that as a blocker anymore.

> We now already have that precedent in the built-in rebase that we implemented a feature only in the C code, and simply forced that in the test case by setting *_USE_BUILTIN=true inside the test case.

Ah nice, I wasn't aware of that. If we have precedent for that, I'll retract my concerns on the project completely.

I'd be willing to co-mentor the project, but as first time mentor I wouldn't feel quite comfortable solo mentoring this or other projects this year.

Reply:
@tgummerer The main point of a mentor is not to know it all, but to help the student, especially in the first weeks when they are likely to be overwhelmed with the project and with little idea how they are supposed to do things.

If there are any questions about details in those projects, I would hope that any capable mentor would guide their interns to the best person(s) to get answers.

Reply:
Right, I'm less worried about the technical side of things than the time commitment necessary.  I raised this concern at Git Merge as well, and Peff and Christian suggested that co-mentoring might be a good option for the first year, so that's what I'd like to stick to for now, and then get a feeling to see how much time I can invest in the future.